### PR TITLE
State: Don't persist current user state after logout, since that hides "email me a login link" functionality

### PR DESCRIPTION
**Pull request by DavidRothstein**
If you log in to Calypso, log out, and then go to the login page (https://wordpress.com/log-in) again, you'll notice the "Email me a login link" (a.k.a. magic links) option is missing from underneath the login form.

What's happening is that after you log out, `state.currentUser` still has the data from your previous session in it, and the code that normally displays the "email me a login link" option consequently thinks you're already logged in.

Based on the way state is persisted, it looks like this bug will continue for around 7 days after logout.

The fact that `state.currentUser` has stale data after logout seems like a pretty fundamental problem, so I wonder what else might be broken by it.  (I encountered this particular bug, though, since we're looking to improve the way magic link logins work for who log in via a billing-related email reminder.)

### CODE REVIEW

**Comment by DavidRothstein** [comment]
The way I fixed the bug here so far doesn't seem very clean, but it does work, and I wasn't sure of what a cleaner way would be.  Ideas for improvement definitely welcome.

**Comment by DavidRothstein** [comment]
This failed the `wp-e2e-tests-full` tests 3 different times, but with completely different test failures each time:
https://circleci.com/gh/Automattic/wp-e2e-tests-for-branches/10073
https://circleci.com/gh/Automattic/wp-e2e-tests-for-branches/10084
https://circleci.com/gh/Automattic/wp-e2e-tests-for-branches/10094

I'm not sure what to make of that.

**Comment by gwwar** [comment]
🤔 we may want to check if `currentUser` is getting unset properly on logout. Another possibility may be to wipe out that key when reading from `redux-state-logged-out` on store init.

**Comment by DavidRothstein** [comment]
Added @ockham for a possible review based on https://github.com/Automattic/wp-calypso/pull/23640 and discussion at p1529527736000239-slack-calypso

File: client/state/initial-state.js
Comment (gwwar) [inline client/state/initial-state.js]: Could we try moving this to `loadInitialState` when we're loading state from `redux-state-logged-out`? I think we can do this one time, unless some user id is being set incorrectly in redux for logged out folks.

Comment (gwwar) [reply client/state/initial-state.js]: Let's also add a test case in `state/test/initial-state.js`

Comment (dmsnell) [reply client/state/initial-state.js]: also how about not using `delete` but instead creating a new copy of the state? notice how _the very next lines_ should return from this function without updating anything because we thought that the two were equal

```js
const prevState = reduxStore.getState();
const { currentUser, ...loggedOutState } = prevState;

const nextState = isLoggedIn()
	? prevState
	: loggedOutState;

if ( state && nextState === state ) {
	return;
}
```

**Comment by jsnajdr** [comment]
The root cause of the issue seems to be: After a successful login ([rebootAfterLogin](https://github.com/Automattic/wp-calypso/blob/28a398ff7ed49b60254c6194e033a04171bee256/client/blocks/login/index.jsx#L120)) or logout ([here](https://github.com/Automattic/wp-calypso/blob/28a398ff7ed49b60254c6194e033a04171bee256/client/me/sidebar/index.jsx#L55) or [here](https://github.com/Automattic/wp-calypso/blob/28a398ff7ed49b60254c6194e033a04171bee256/client/me/reauth-required/index.jsx#L206)) we clear the local storage and then redirect by calling:
```js
user.clear( () => window.location.href = rebootUrl );
```
The `user.clear` call deletes all user data from local storage (for both the logged-out -> logged-in and logged-in -> logged-out transitions) and then reboots Calypso by navigating to another URL.

However, the Redux state remains in memory. That is not a problem, because the memory state will be wiped out on the app reboot. But it becomes a problem when it is persisted in the [`window.beforeunload` event handler](https://github.com/Automattic/wp-calypso/blob/d8a9677e35e89e0be02a7b912b51d2058ee58f01/client/state/initial-state.js#L136).

Right after it was removed from local storage, it is persisted again! The sequence of events is:
1. call `user.clear` and clear local storage
2. set `window.location.href` to a new URL
3. that triggers `beforeunload` and the state that is still in memory is persisted again
4. only now the browser navigates to the new URL, shutting down the old instance of the app and booting a new one.

How to solve that? Persisting the state on `beforeunload` is legitimate when log-in or log-out is not in progress -- we want the latest data to be persisted when simply closing the browser tab. So we can't just remove the `beforeunload` handler. We only want to disable it when app reboot is in progress.

A good solution seems to be that the `user.clear` function disables the state persistence:
```js
import { disablePersistence } from 'state/initial-state';

User.prototype.clear = () => {
  ...
  disablePersistence();
  ...
};
```

`user.clear` is always called immediately before app reboot, so it can successfully and correctly make the `beforeunload` a noop.

The current patch that deletes the `currentUser` from Redux state is merely scratching the surface: all other private data (posts etc.) are still stored to local storage, although it should be clear. After I logout from a public computer, my private data are still there. :-1:

**Comment by DavidRothstein** [comment]
Thanks for all the reviews.  I have some code to switch to the method that @gwwar recommended, and also started writing some tests.  But I'll put that on hold while we figure out the correct implementation.

@jsnajdr What you're saying makes a lot of sense, but does it work in the case where the user logs out via another method (besides Calypso)?  For example, I could be mistaken, but I thought that if you're on `yoursite.example.com` and use the "Sign Out" option there, that won't trigger any Calypso logout code.

Either way, what you're saying does seem like it's part of the solution.  You can see from the code comment I wrote that I anticipated this would eventually need to be done for more than just the `currentUser` property... But I wasn't sure if it was intentional that some state is persisted after logout.  Given all the `user.clear()` calls you mentioned, it sounds like it isn't.

So *if* the issue I mentioned above is a real one (logouts ouside of Calypso), a possible combined solution to handle all cases would be to check in `loadInitialState` (similar to @gwwar's recommendation), but if we see the user is logged out while `state.currentUser` still indicates the user is logged in, clear *all* state then (not just the `currentUser` property)?  This could maybe even be combined with the idea of disabling persistence on `user.clear()` - to handle the most common logout cases before the bug even happens.

File: client/state/initial-state.js
Comment (DavidRothstein) [reply client/state/initial-state.js]: > also how about not using delete but instead creating a new copy of the state?

Good point - we probably shouldn't be modifying the main state object like that.

(However, now that we're looking to move to a different implementation, this issue probably won't come up again there.)

**Comment by DavidRothstein** [comment]
> @jsnajdr What you're saying makes a lot of sense, but does it work in the case where the user logs out via another method (besides Calypso)? For example, I could be mistaken, but I thought that if you're on yoursite.example.com and use the "Sign Out" option there, that won't trigger any Calypso logout code.

Actually I guess it's mostly OK - since the Calypso logout code is *also* what causes the bug in the first place. So if you log out without triggering the Calypso logout code, you won't trigger the bug either :)

So doing it via `user.clear()` alone could be OK after all, at least for the purposes of this bug fix (it wouldn't be guaranteed to fix it for anyone *already* experiencing the bug, but that should resolve itself in 7 days when the stored state expires).

Not sure it fully solves this though:

> After I logout from a public computer, my private data are still there

Since if you log out from a public computer but don't go through the Calypso logout flow, your private data may still be in the local store.

**Comment by gwwar** [comment]
Also summoning @ockham, @bisko or another representative from @Automattic/jetpack be able to chat a bit about the desired connection workflow for logged out folks from https://github.com/Automattic/wp-calypso/pull/23640 to make sure that we're not breaking existing flows.

I personally find that not persisting state for logged-out folks is simpler to reason about (see past behavior in https://github.com/Automattic/wp-calypso/pull/13185). Since state can be persisted in other ways more reliably, such as in the session or in the url itself for connection flows, where we don't need to worry about having browser support for this.

> After I logout from a public computer, my private data are still there

@jsnajdr to clarify is this referring to `localForage` instance (which uses many drivers) or our old localStorage wrapper eg `require( 'store' )`?

**Comment by bisko** [comment]
Note: I’m referring only to the main part of Signup, not Jetpack flows as I haven’t dug into them to know how it affects them.

The reason for persisting Signup state was because how the framework was handling the state. 

The data store architecture was split between two “stores” - one was migrated to Redux, while the other remained in Flux. The Flux store was persisting information to `localStorage` which in turn meant that when Signup was initialized for logged-out it was loading only half of the data - the `localStorage` one, which was pushed in the Flux store. 

That as you might imagine was causing some broken behaviors and in the end resulted broken ( blank page ) signup if you reloaded the page mid-site-creation and being logged-out.

With the new improvements @jsnmoon is working on to Reduxify Signup, this won’t be a problem anymore and it wouldn’t be an issue to not persist state.

It will still be a bit of degraded experience as you’ll have to start Signup from scratch, but since it’s a short process it won’t be that much of an issue.

**Comment by gwwar** [comment]
Thanks for sharing the extra context @bisko ✨

**Review by ockham — CHANGES_REQUESTED** [review]
Whatever we do, let's please not special-case removal of `state.currentUser` only. Let's either keep all of `state`, or wipe it out. Persisting only parts of the state is bound to get us in trouble at some point.

I'll mostly defer to @jsnajdr analysis and suggestions -- TBH, I wasn't even aware of all that's going on upon logout.

I'd love to keep Redux state persistence for logged-out users, for cases such as Jetpack Onboarding where we can cut some corners by persisting when redirecting back and forth between (logged-out) Calypso and wp-admin. I know our stance is to see Redux state persistence as an enhancement and not something to rely on simply because we've deemed invocation of that `beforeUnload` handler too unreliable (cc @gwwar), but reading Jarda's description of what is invoked makes me wonder if that unreliability might really be due to race conditions arising out of that complexity? Also, I've never really liked the idea of resorting to other persistence mechanisms (such as `localForage` or whatnot). Using a different library here just feels like a dirty workaround to an issue we should probably fix differently. /cc @mtias @dmsnell 

That said, it makes sense that we would want logged-out Calypso to be ignorant of Redux state created while logged-in -- but I'm somewhat puzzled this isn't the case already, since the whole point of #23640 was to avoid pollution across logged-in vs logged-out state by using keys (user IDs while logged in vs the `logged-out` string while logged out, [see](https://github.com/Automattic/wp-calypso/pull/23640/files#diff-e18463a8ac59a4ed5a9b2f702e4cc4a3R108)). @DavidRothstein Can you track down why/how this mechanism fails?

**Comment by jsnajdr** [comment]
> Can you track down why/how this mechanism fails?

@ockham I described that in my comment above:
1. During logout, `User.prototype.clear` clears the local storage and the `User` object, Redux state remains in memory
2. Page is reloaded in order to reboot Calypso
3. There is `beforeunload` handler that saves the Redux state again. This time as `redux-state-logged-out`, because `User` is empty already.

> to clarify is this referring to localForage instance (which uses many drivers) or our old localStorage wrapper eg require( 'store' )?

@gwwar `User.prototype.clear` clears both of them.

> Since if you log out from a public computer but don't go through the Calypso logout flow, your private data may still be in the local store.

@DavidRothstein Yes, that's true. When logging out using the WP.com masterbar, local storage is not cleared. In that case, my private data remain on the browser unless the session was in a private window.

Inside Calypso, all logout flows clear the local storage though. The fact that the `beforeunload` handler sabotages this effort is an obvious bug.

(I wrote a reply similar to the above yesterday, but now I see that it's not here? Was it lost on the network or deleted? 😕 )

**Comment by DavidRothstein** [comment]
> the whole point of #23640 was to avoid pollution across logged-in vs logged-out state by using keys (user IDs while logged in vs the logged-out string while logged out) .... @DavidRothstein Can you track down why/how this mechanism fails?

I don't have a whole lot to add beyond what @jsnajdr wrote above: It's basically because the logged-in user's Redux state is still in memory after logout, so that's what gets written to the `logged-out` store on `beforeunload`.

My very first attempt to deal with this (before even posting the PR) was actually just to try to erase the stale data from the Redux state when the user logs out.  (https://stackoverflow.com/questions/35622588/how-to-reset-the-state-of-a-redux-store/35641992#35641992 might be a good way to do that, although it's not what I originally tried.)  But that led to lots of errors during logout, since it seems like some parts of the request still relied on the original state being there.  So I gave up on it, but maybe it's possible to get it to work.

Otherwise, if the goal is to keep state persistence for logged-out users in place, then I guess the options now are:
- Make `user.clear()` force persistence to be disabled (@jsnajdr's solution).
- Possibly a simpler solution within `persistOnChange` itself - just compare `state.currentUser.id` to `user.get().ID` and only save the state if they match?

**Comment by dmsnell** [comment]
@DavidRothstein it sounds like our course of action should be to close this PR, create a new one and reference this one, create a new description, and try out either (1) or (2) in your previous comment.

Feel free to spend a small amount of time exploring both and then summarize your findings: why prefer one over the other, or maybe there's no real preference.

The arguments here have made it sound like working around the issue at hand is going to cause further pain down the line.

**Comment by DavidRothstein** [comment]
Sounds good.  I'll probably start with (2) since it seems a little more self-contained.  I might not get a chance to look at this until next week though.

**Comment by ockham** [comment]
>> Can you track down why/how this mechanism fails?

> @ockham I described that in my comment above: [...]

@jsnajdr Right, sorry about that. I was in a bit of a rush when posting my comment. I realized this was a silly question after I'd posted it since you'd already answered it.

**Comment by DavidRothstein** [comment]
New pull request now posted here: https://github.com/Automattic/wp-calypso/pull/25907

Thanks for the help here, everyone!  Please head over there if you want to review the new proposed fix.

**Pull request closed without merging**
